src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/DynamicMethod.cs

```diff
@@ -431,36 +465,126 @@ internal RuntimeMethodHandle GetMethodDescriptor()
                 throw new TargetParameterCountException(SR.Arg_ParmCnt);
 
-            // if we are here we passed all the previous checks. Time to look at the arguments
-            bool wrapExceptions = (invokeAttr & BindingFlags.DoNotWrapExceptions) == 0;
+            object? retValue;
 
-            StackAllocedArguments stackArgs = default;
-            Span<object?> arguments = default;
-            if (actualCount != 0)
+            unsafe
             {
-                arguments = CheckArguments(ref stackArgs, parameters, binder, invokeAttr, culture, sig.Arguments);
+                if (argCount == 0)
+                {
+                    retValue = Invoker.InvokeUnsafe(obj, args: default, invokeAttr);
+                }
+                else if (argCount > MaxStackAllocArgCount)
+                {
+                    Debug.Assert(parameters != null);
+                    CheckManyArguments();
+                }
+                else
+                {
+                    Debug.Assert(parameters != null);
+                    StackAllocedArguments argStorage = default;
+                    Span<object?> copyOfParameters = new Span<object?>(ref argStorage._arg0, argCount);
+                    Span<bool> shouldCopyBackParameters = new Span<bool>(ref argStorage._copyBack0, argCount);
+
+                    StackAllocatedByRefs byrefStorage = default;
+                    IntPtr* pByRefStorage = (IntPtr*)&byrefStorage;
+
+                    CheckArguments(
+                        copyOfParameters,
+                        pByRefStorage,
+                        shouldCopyBackParameters,
+                        parameters,
+                        Signature.Arguments,
+                        binder,
+                        culture,
+                        invokeAttr);
+
+                    retValue = Invoker.InvokeUnsafe(obj, pByRefStorage, invokeAttr);
+
+                    // Copy modified values out. This should be done only with ByRef or Type.Missing parameters.
+                    for (int i = 0; i < argCount; i++)
+                    {
+                        if (shouldCopyBackParameters[i])
+                        {
+                            parameters[i] = copyOfParameters[i];
+                        }
+                    }
+                }
             }
 
-            object? retValue = RuntimeMethodHandle.InvokeMethod(null, arguments, sig, false, wrapExceptions);
-
-            // copy out. This should be made only if ByRef are present.
-            // n.b. cannot use Span<T>.CopyTo, as parameters.GetType() might not actually be typeof(object[])
-            for (int index = 0; index < arguments.Length; index++)
-                parameters![index] = arguments[index];
-
             GC.KeepAlive(this);
             return retValue;
+
+            // Slower path that does a heap alloc for copyOfParameters and registers byrefs to those objects.
+            unsafe void CheckManyArguments()
```

Comment:
I think this needs to be `static` and all arguments passed explicitly. Otherwise, the compiler will create a helper struct to spill all shared arguments and locals into that tends to hurt perf.

Reply:
OK; I'll update the perf numbers after a bit.